### PR TITLE
[tests-only][full-ci] add tests for CLI command uploads sessions

### PR DESCRIPTION
**Pull request by PrajwolAmatya**
## Description
This PR adds test for CLI command related to uploads sessions.

Test scenario covered:
```feature
Scenario: user lists all upload sessions
Scenario: user lists all upload sessions that are currently postprocessing
Scenario: user lists all upload sessions that are infected by virus
Scenario: user lists all expired upload sessions
Scenario: user cleans all expired upload sessions
```

## Related Issue

- https://github.com/owncloud/ocis/issues/10167

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Review by phil-davis — APPROVED** [review]
LGTM
Note: we could write "Then" steps in the form "Then the CLI response should contain only these entries". And that step would check for the required entries, and also check that there are no other entries. And then we would not need the "should not contain" steps.

But it reads nicely like it is - the scenarios are clear about what they are checking for.

File: tests/acceptance/bootstrap/CliContext.php [hunk 1]
```diff
@@ -233,4 +234,88 @@ public function theCommandOutputShouldContain(string $shouldOrNot, string $outpu
...
+	public function theAdministratorCleansAllExpiredUploadSessions(): void {
+		$command = "storage-users uploads sessions --expired --clean --json";
+		$body = [
+			"command" => $command
+		];
+		$this->featureContext->setResponse(CliHelper::runCommand($body));
+	}
+
+	/**
+	 * @Then /^the CLI response should (not|)\s?contain these entries:$/
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/CliContext.php:271]: should be more readable
```suggestion
	 * @Then /^the CLI response (should|should not) contain these entries:$/
```

File: tests/acceptance/features/cliCommands/uploadSessions.feature [hunk 2]
```diff
@@ -0,0 +1,83 @@
+@env-config @uploads-sessions
```
Comment (saw-jan) [inline tests/acceptance/features/cliCommands/uploadSessions.feature:1]: maybe we should mention cli
```suggestion
@env-config @cli-uploads-sessions
```

File: tests/acceptance/bootstrap/CliContext.php [hunk 3]
```diff
@@ -233,4 +234,88 @@ public function theCommandOutputShouldContain(string $shouldOrNot, string $outpu
...
+	 * @param string $shouldOrNot (not|)
+	 * @param TableNode $table
+	 *
+	 * @return void
+	 */
+	public function theCLIResponseShouldContainTheseEntries(string $shouldOrNot, TableNode $table): void {
+		$expectedFiles = $table->getColumn(0);
+		$responseBody = $this->featureContext->getJsonDecodedResponse($this->featureContext->getResponse());
+		preg_match('/(\[.*\])/', $responseBody["message"], $matches);
+		$responseArray = \json_decode($matches[1], null, 512, JSON_THROW_ON_ERROR);
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/CliContext.php:282]: some comment with example here would be great to have
```suggestion
		preg_match('/(\[.*\])/', $responseBody["message"], $matches);
		$responseArray = \json_decode($matches[1], null, 512, JSON_THROW_ON_ERROR);
```

File: tests/acceptance/bootstrap/CliContext.php [hunk 4]
```diff
@@ -233,4 +234,88 @@ public function theCommandOutputShouldContain(string $shouldOrNot, string $outpu
...
+			}
+		}
+	}
+
+	/**
+	 * @AfterScenario @uploads-sessions
+	 *
+	 * @return void
+	 */
+	public function cleanUploadsSessions(): void {
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/CliContext.php:313]: hooks must be static
```suggestion
	public static function cleanUploadsSessions(): void {
```

File: tests/acceptance/bootstrap/CliContext.php [hunk 5]
```diff
@@ -233,4 +234,92 @@ public function theCommandOutputShouldContain(string $shouldOrNot, string $outpu
...
+				);
+			}
+		} else {
+			foreach ($expectedFiles as $expectedFile) {
+				Assert::assertTrue(
+					\in_array($expectedFile, $resourceNames),
+					"The resource '$expectedFile' was not found in the response."
+				);
+			}
+		}
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/CliContext.php:309]: though the end result will be the same.
```suggestion
		foreach ($expectedFiles as $expectedFile) {
			if ($shouldOrNot === "should not") {
				Assert::assertNotTrue(
					\in_array($expectedFile, $resourceNames),
					"The resource '$expectedFile' was found in the response."
				);
			} else {
				Assert::assertTrue(
					\in_array($expectedFile, $resourceNames),
					"The resource '$expectedFile' was not found in the response."
				);
			}
		}
```

File: tests/acceptance/bootstrap/CliContext.php [hunk 6]
```diff
@@ -233,4 +234,92 @@ public function theCommandOutputShouldContain(string $shouldOrNot, string $outpu
...
+	 *
+	 * @return void
+	 */
+	public static function cleanUploadsSessions(): void {
+		$command = "storage-users uploads sessions --clean";
+		$body = [
+			"command" => $command
+		];
+		$response = CliHelper::runCommand($body);
+		Assert::assertEquals("200", $response->getStatusCode());
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/CliContext.php:323]:
```suggestion
		Assert::assertEquals("200", $response->getStatusCode(), "Fail to clean upload sessions");
```

File: tests/acceptance/bootstrap/CliContext.php [hunk 7]
```diff
@@ -233,4 +234,92 @@ public function theCommandOutputShouldContain(string $shouldOrNot, string $outpu
...
+			}
+		}
+	}
+
+	/**
+	 * @AfterScenario @cli-uploads-sessions
+	 *
+	 * @return void
+	 */
+	public static function cleanUploadsSessions(): void {
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/CliContext.php:317]: my bad. this is a `scenario` hook, so this should not necessarily be a static method.https://docs.behat.org/en/v2.5/guides/3.hooks.html#scenario-hooks

>hooks must be static. See https://docs.behat.org/en/v2.5/guides/3.hooks.html#suite-hooks

`suite` and `feature` hooks must be static.

Soryy, for the confusion :cry:

File: tests/acceptance/bootstrap/CliContext.php (hunk 5 above)
Comment (PrajwolAmatya) [reply tests/acceptance/bootstrap/CliContext.php:309]: Wouldn't this check the if condition on every loop? Checking the condition only once for one particular step would be good IMO.

Comment (saw-jan) [reply tests/acceptance/bootstrap/CliContext.php:309]: let's keep it that way :+1:

**Pull request merged**